rfcs/convergence/simplify-prop-merging.md

```diff
@@ -0,0 +1,358 @@
+      /** Can be a rocket 🚀 or a footgun 🔫 */
+    },
+  }}
+/>;
+```
+
+### Retire mergeProps 🙊
+
+Since internally even the Fluent team is mutating state, there is no clear reason why we should prevent consumers from doing so with `defaultProps`. In fact even if consumers do use `defaultProps` there is still nothing that **stops** them from mutating state. We show in the following example, two ways of doing the exact same thing. Both options are currently equally possible and valid.
+
+```tsx
+const state = mergeProps(
+  {
+    foo: {},
+  },
+  defaultProps,
+  resolveShorthandProps(props, buttonShorthandProps),
+);
+
+// Option 1 - defaultProps
+useComponentState(props, { foo: { bar: 'bar' } });
+
+// Option 2 - mutate state
+const state = useComponentState(props);
+state.foo.bar = 'bar';
+```
+
+With this statement in mind, we can propose solution with `mergeProps`. Let's also simplify shorthand resolution to resolve shorthands for each slot separately. Let's also apply the previous solutions to see the final before/after.
+
+#### Current usage of `mergeProps()`
+
+```tsx
+const state = mergeProps(
+  {
+    ref: resolvedRef,
+    as: 'button',
+    icon: { as: 'span' },
+    content: { as: 'span', children: props.children },
+    loader: { as: Loader, role: 'presentation' },
+  },
+  defaultProps,
+  resolveShorthandProps(props, buttonShorthandProps),
+);
+```
+
+#### Proposed usage
+
+```tsx
+const state: FooState = {
+  as: "button",
+
+  components: {
+    icon: "span",
+    content: "span",
+    loader: Loader,
+  },
+
+  // user props
+  ...props,
+
+  ref: resolvedRef,
+  content: resolveShorthand(
+    props.content,
+    { children: props.children } /* defaultProps for a slot */
+  ),
+  icon: resolveShorthand(props.icon),
+  loader: resolveShorthand(
+    props.loader,
+    { role: "presentation" } /* defaultProps for a slot */
+  ),
+};
+
+export interface FooProps extends ComponentProps, React.ButtonHTMLAttributes<HTMLButtonElement> {
+  icon?: ShorthandProps<React.HTMLAttributes<HTMLSpanElement>;
+  content?: ShorthandProps<React.HTMLAttributes<HTMLSpanElement>;
+  Loader?: ShorthandProps<LoaderProps>;
+};
+
+export interface FooState extends FooProps, React.ButtonHTMLAttributes<HTMLButtonElement> {
+  icon: ObjectShorthandProps<React.HTMLAttributes<HTMLSpanElement>;
+  content: ObjectShorthandProps<React.HTMLAttributes<HTMLSpanElement>;
+  Loader: ObjectShorthandProps<LoaderProps>;
+};
+```
+
+Typings there are 100% clear since they directly map to types 🎉
+
+Now when either internally or externally we want to extend/add slots:
+
+```ts
+const state = useComponentState();
+state.components.extraSlot = 'div';
+state.extraSlot = resolveShorthand({
+  /** Shorthand props */
+});
+
+state.components.icon = 'button';
+state.icon = filterNativeButtonProps(state.icon);
+```
+
+This results in use of standard object spread and explicit resolution of slots in state rather than a deep merging all user and default props.
```

Comment:
```suggestion
This results in the use of standard object spread and explicit resolution of slots in state, rather than deep merging all user and default props.
```